### ISSUE TITLE
pageserver: add "critical events" counter for highly impactful error paths

### ISSUE THREAD

**Issue by jcsp** [neondatabase/neon#10094]
There are certain errors that are intrinsically "scary" things that we always want to know about right away:
- walredo failures
- getpage requests that can't find a key
- WAL records we can't ingest
- tenants we can't load (broken tenants).
- 404 loading a layer file that the index thinks should exist (this _can_ occur legitimately if an isolated PS is still attached to something attached elsewhere, but that is super rare)

These are mostly detectable some other way, but we have a plethora of different benchmarks, tests and deployed environments, and we need a simple single-metric thing that anyone can query to detect an unambiguous "You have hit a serious storage (_maybe_ compute but probably storage) bug".

**Comment by skyzh** [comment]
> getpage requests that can't find a key

the same applies to GC/compaction, we should match on all `could not find key`, not only on the read path

**Comment by erikgrinaker** [comment]
Why is there no `tracing::critical!`? :disappointed: Seems like a log level would be more appropriate than matching on random error messages (of course, we'd still have to tag the relevant errors as such).

**Comment by jcsp** [comment]
>Seems like a log level would be more appropriate than matching on random error messages (of course, we'd still have to tag the relevant errors as such).

Yeah, I also wish `tracing` had built in counters for messages of each severity.  

I expect to add some global `critical_event(&str)` function that logs at ERROR and increments the counter.  Alerting should be driven by the metric so that we don't have to go add log-driven metrics for each case.

**Comment by Bodobolero** [comment]
once we have this task implemented pls create additional issues
a) create a GitHub action that a workflow/testcase can easily add to the testcase/job to validate the metric has the expected value
b) create a Git issue for each benchmark/testcase owner that you think should add instrumentation for checking the  metric

**Comment by jcsp** [comment]
@bayandin can devprod take ownership of the parts about benchmarks + github actions that Peter mentions above?

**Comment by erikgrinaker** [comment]
> Yeah, I also wish `tracing` had built in counters for messages of each severity.

`tracing` may not have it, but we've plugged it in as `libmetrics_tracing_event_count`:

https://github.com/neondatabase/neon/blob/2d885ac07ae0207ab886fd4dda84701ae33893f1/libs/utils/src/logging.rs#L36-L44

I'm going to add in a `critical` level for this metric, along with a `critical!` macro.

**Comment by erikgrinaker** [comment]
I added a `critical!` macro in #10641 along with a global platform oncall alerting rule (both production and staging).

I'll leave this open to migrate some appropriate errors and replace existing log alerting rules.

**Comment by erikgrinaker** [comment]
This has been deployed now. I'm going to track the remaining operational tasks separately.
